src/cms/logic.py

```diff
@@ -0,0 +1,323 @@
+    for excluded_url in excluded_urls():
+        if excluded_url in deeper_url:
+            return False
+    return True
+
+
+def remove_fragment(url):
+    fragment = urlparse(url).fragment
+    return url.replace('#' + fragment, '')
+
+
+def url_is_unique(deeper_url):
+    def normalize(url):
+        url = remove_fragment(url)
+        if url.endswith('/'):
+            url = url[:-1]
+        return url
+
+    if deeper_url in fetched_urls:
+        return False
+    for fetched_url in fetched_urls:
+        if normalize(fetched_url) == normalize(deeper_url):
+            return False
+    return True
+
+
+def decompose_non_content_page_regions(body):
+    for non_content_selector in ['header', 'footer', 'h1']:
+        for element in body.select(non_content_selector):
+            element.decompose()
+
+
+def add_searchable_page(url):
+
+    html = get_page(url)
+    if not html:
+        return
+
+    body = get_body(html)
+    if not body:
+        return
+
+    for anchor in body.find_all('a'):
+        href = anchor.get('href', '').strip()
+        deeper_url = urljoin(url, href)
+        if url_in_scope(deeper_url) and url_is_unique(deeper_url):
+            add_searchable_page(deeper_url)
+
+    name = get_name(html)
+    decompose_non_content_page_regions(body)
+
+    add_searchable_page_parts(url, body)
+    add_search_index_document(url, name, body.get_text())
+
+
+def get_press_site_search_data():
+    """
+    Generates data for press-level site search index
+    to be used by MiniSearch
+    """
+
+    base = get_base()
+    add_searchable_page(base)
+
+    if not len(search_documents) > 0:
+        logger.error('Search data store is empty')
+
+    return search_documents
+
+
+def update_search_data(press_id=1):
+    press = press_models.Press.objects.get(pk=press_id)
+    docs_filename = os.path.join(
+        settings.SITE_SEARCH_DIR,
+        f'_press_{ press.pk }_documents.json'
+    )
+    docs_file, created = models.MediaFile.objects.get_or_create(
+        label=docs_filename
+    )
+    if not created:
+        docs_file.unlink()
+
+    documents = get_press_site_search_data()
+    docs_json = json.dumps(documents, separators=(',', ':'))
+
+    content_file = ContentFile(docs_json.encode('utf-8'))
+    docs_file.file.save(docs_filename, content_file, save=True)
+    return docs_file
+
+
+def delete_search_data(press_id=1):
+    press = press_models.Press.objects.get(pk=press_id)
+    files_deleted = []
+    path = os.path.join(
+        SITE_SEARCH_PATH,
+        f'_press_{ press.pk }_documents.json',
+    )
+    if os.path.exists(path):
+        os.unlink(path)
+        files_deleted.append(path)
+    if settings.IN_TEST_RUNNER:
+        if os.listdir(SITE_SEARCH_PATH):
+            logger.warning(
+                f'Left-over test files: {os.listdir(SITE_SEARCH_PATH)}'
+            )
+        else:
+            os.rmdir(SITE_SEARCH_PATH)
+
+    return files_deleted
+
+
+@cache(900)
+def get_search_data_file(press):
```

Comment:
If we turn this into `get_search_data_file` into `get_search_data_url` and return instead `docs_file.url` we also save us from having to run the query on every request.